usecases/webhook_events_usecase.go

```diff
@@ -108,102 +141,98 @@ func (usecase WebhookEventsUsecase) SendWebhookEvents(
 		return nil
 	}
 
-	var waitGroup sync.WaitGroup
-	// The channel needs to be big enough to store any possible errors to avoid deadlock due to the presence of a waitGroup
-	uploadErrorChan := make(chan error, len(pendingWebhookEvents))
-	deliveryStatusChan := make(chan models.WebhookEventDeliveryStatus, len(pendingWebhookEvents))
+	group, ctx := errgroup.WithContext(ctx)
+	group.SetLimit(MAX_CONCURRENT_WEBHOOKS_SENT)
 
-	startProcessSendWebhookEvent := func(webhookEventId string) {
-		defer waitGroup.Done()
-		logger := logger.With("webhook_event_id", webhookEventId)
-		deliveryStatus, err := usecase.sendWebhookEvent(ctx, webhookEventId, logger)
-		if err != nil {
-			uploadErrorChan <- err
-		}
-		if deliveryStatus != nil {
-			deliveryStatusChan <- *deliveryStatus
-		}
-	}
+	deliveryStatuses := make([]models.WebhookEventDeliveryStatus, len(pendingWebhookEvents))
 
-	for _, webhookEvent := range pendingWebhookEvents {
-		waitGroup.Add(1)
-		go startProcessSendWebhookEvent(webhookEvent.Id)
-	}
+	for i, webhookEvent := range pendingWebhookEvents {
+		group.Go(func() error {
+			ctx := utils.StoreLoggerInContext(
+				ctx,
+				logger.With("webhook_event_id", webhookEvent.Id))
+
+			select {
+			case <-ctx.Done():
+				return errors.Wrapf(ctx.Err(), "context cancelled before retrying webhook %s", webhookEvent.Id)
+			default:
+			}
 
-	waitGroup.Wait()
-	close(uploadErrorChan)
-	close(deliveryStatusChan)
+			deliveryStatus, err := usecase._sendWebhookEvent(ctx, webhookEvent.Id)
+			deliveryStatuses[i] = deliveryStatus
```

Comment:
Same here, we may use some "invalid" deliveryStatus

Reply:
Well, here in fact we don't, at least we don't use it: any error returned will return an error to the waitgroup (line 167 `err = group.Wait()`) after which we don't use the deliveryStatuses slice.
Still I get your point.